node/pkg/driver/node_test.go

```diff
@@ -895,23 +896,141 @@ func TestNodeUnpublishVolume(t *testing.T) {
 }
 
 func TestNodeGetVolumeStats(t *testing.T) {
+	volId := "someStorageType:vol-test"
+	volumePath := "/test/path"
+	stagingTargetPath := "/staging/test/path"
+	volumePathWithHostPrefix := GetPodPath(volumePath)
 
-	req := &csi.NodeGetVolumeStatsRequest{}
+	testCases := []struct {
+		name     string
+		testFunc func(t *testing.T)
+	}{
+		{
+			name: "fail volumePath does not exists",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+				mockNodeUtils := mocks.NewMockNodeUtilsInterface(mockCtl)
+				d := newTestNodeService(mockNodeUtils, nil, nil)
+				expErrCode := codes.NotFound
+				mockNodeUtils.EXPECT().IsPathExists(volumePathWithHostPrefix).Return(false)
+				mockNodeUtils.EXPECT().GetPodPath(volumePath).Return(volumePathWithHostPrefix)
+				req := &csi.NodeGetVolumeStatsRequest{
+					VolumeId:          volId,
+					VolumePath:        volumePath,
+					StagingTargetPath: stagingTargetPath,
+				}
 
-	d := newTestNodeService(nil, nil, nil)
+				_, err := d.NodeGetVolumeStats(context.TODO(), req)
+				if err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d error code, message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+				} else {
+					t.Fatalf("Expected to get an error, got nil")
+				}
+			},
+		},
+		{
+			name: "fail to get stats",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+				mockNodeUtils := mocks.NewMockNodeUtilsInterface(mockCtl)
+				d := newTestNodeService(mockNodeUtils, nil, nil)
+				mockNodeUtils.EXPECT().GetPodPath(volumePath).Return(volumePathWithHostPrefix)
+				mockNodeUtils.EXPECT().IsPathExists(volumePathWithHostPrefix).Return(true)
+				mockNodeUtils.EXPECT().IsBlock(volumePathWithHostPrefix).Return(false, nil)
+				mockNodeUtils.EXPECT().IsNotMountPoint(volumePathWithHostPrefix).Return(false, nil)
+				mockNodeUtils.EXPECT().IsDirectory(volumePathWithHostPrefix).Return(true)
+				mockNodeUtils.EXPECT().GetVolumeStats(volumePathWithHostPrefix).Return(driver.VolumeStatistics{}, status.Errorf(codes.Internal, "fail to get stats"))
+				req := &csi.NodeGetVolumeStatsRequest{
+					VolumeId:          volId,
+					VolumePath:        volumePath,
+					StagingTargetPath: stagingTargetPath,
+				}
 
-	expErrCode := codes.Unimplemented
+				_, err := d.NodeGetVolumeStats(context.TODO(), req)
+				if err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if !strings.Contains(err.Error(), "Failed to get capacity statistics") {
+						t.Fatalf("Expected to get error from GetVolumeStats function, got error message: %s", err.Error())
+					}
+				} else {
+					t.Fatalf("Expected to get an error, got nil")
```

Comment:
this looks like code dup.
would moving the actual `NodeGetVolumeStats` calls into a common loop at the end of the function, make the code shorter/cleaner (as was done e.g. [here](https://github.com/IBM/ibm-block-csi-driver/pull/450))?

Reply:
I moved some calls to functions (not all the `NodeGetVolumeStats` calls), I think it's cleaner, is it enough? or there is something else I can do here?

Reply:
much better